File: src/test/java/com/angorasix/projects/core/infrastructure/persistence/converter/ZonedDateTimeConvertersUnitTest.java
package com.angorasix.projects.core.infrastructure.persistence.converter;

import static org.assertj.core.api.Assertions.assertThat;

import com.angorasix.projects.core.infrastructure.persistence.converter.ZonedDateTimeConverters.ZoneDateTimeReaderConverter;
import com.angorasix.projects.core.infrastructure.persistence.converter.ZonedDateTimeConverters.ZoneDateTimeWritingConverter;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import org.bson.Document;
import org.junit.jupiter.api.Test;

public class ZonedDateTimeConvertersUnitTest {

  final ZoneDateTimeWritingConverter writingConverter = new ZoneDateTimeWritingConverter();
  final ZoneDateTimeReaderConverter readerConverter = new ZoneDateTimeReaderConverter();

  @Test
  public void givenZonedDateTime_whenWritingToDocument_thenCustomFieldsCorrectlyCreated()
      throws Exception {
    final ZonedDateTime inputZdt =
        ZonedDateTime.of(1991, 04, 22, 23, 00, 00, 0, ZoneId.of("America/Argentina/Cordoba"));

    final Document outputDocument = writingConverter.convert(inputZdt);

    assertThat(outputDocument.get("zone")).isEqualTo("America/Argentina/Cordoba");
    assertThat(outputDocument.getDate("dateTime"))
        .isEqualTo(Date.from(inputZdt.withZoneSameInstant(ZoneId.of("UTC")).toInstant()));
  }

  @Test
  public void asdgivenZonedDateTime_whenWritingToDocument_thenCustomFieldsCorrectlyCreated()
      throws Exception {
    final ZonedDateTime inputZdt =
        ZonedDateTime.of(1991, 04, 22, 15, 00, 00, 0, ZoneId.of("Pacific/Auckland"));

    final Document outputDocument = writingConverter.convert(inputZdt);

    assertThat(outputDocument.get("zone")).isEqualTo("Pacific/Auckland");
    assertThat(outputDocument.getDate("dateTime"))
        .isEqualTo(Date.from(inputZdt.withZoneSameInstant(ZoneId.of("UTC")).toInstant()));
  }

  @Test
  public void givenDocumentWithCorrectFields_whenReadingDocument_thenZonedDateTimeCorrectlyCreated()
      throws Exception {
    final ZonedDateTime baseZdt =
        ZonedDateTime.of(1991, 04, 22, 23, 00, 00, 0, ZoneId.of("America/Argentina/Cordoba"));
    final Map<String, Object> inputDocumentMap = new HashMap<>();
    inputDocumentMap.put("zone", "America/Argentina/Cordoba");
    inputDocumentMap.put("dateTime",
        Date.from(baseZdt.withZoneSameInstant(ZoneId.of("UTC")).toInstant()));
    final Document inputDocument = new Document(inputDocumentMap);

    ZonedDateTime outputTdz = readerConverter.convert(inputDocument);

    assertThat(outputTdz).isEqualTo(baseZdt);
    assertThat(outputTdz.toString()).isEqualTo("1991-04-22T23:00-04:00[America/Argentina/Cordoba]");
  }
}

File: src/test/kotlin/com/angorasix/projects/core/infrastructure/persistence/converter/ZonedDateTimeConvertersUnitTest.kt
package com.angorasix.projects.core.infrastructure.persistence.converter

import org.assertj.core.api.Assertions
import org.assertj.core.api.AssertionsForClassTypes.assertThat
import org.bson.Document
import org.junit.jupiter.api.Test
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.*

class ZonedDateTimeConvertersUnitTest {
    private val writingConverter = ZonedDateTimeConvertersUtils.ZonedDateTimeWritingConverter()
    private val readerConverter = ZonedDateTimeConvertersUtils.ZonedDateTimeReaderConverter()

    // Just for coverage purposes
    val converters = ZonedDateTimeConvertersUtils

    @Test
    @Throws(Exception::class)
    fun givenZonedDateTime_whenWritingToDocument_thenCustomFieldsCorrectlyCreated() {
        val inputZdt = ZonedDateTime.of(
            1991,
            4,
            22,
            23,
            0,
            0,
            0,
            ZoneId.of("America/Argentina/Cordoba"),
        )
        val outputDocument = writingConverter.convert(inputZdt)
        assertThat<Any>(outputDocument!!["zone"]).isEqualTo("America/Argentina/Cordoba")
        Assertions.assertThat(outputDocument.getDate("dateTime"))
            .isEqualTo(
                Date.from(
                    inputZdt.withZoneSameInstant(
                        ZoneId.of("UTC"),
                    )
                        .toInstant(),
                ),
            )
    }

    @Test
    @Throws(Exception::class)
    fun givenDifferentZonedDateTime_whenWritingToDocument_thenCustomFieldsCorrectlyCreated() {
        val inputZdt = ZonedDateTime.of(
            1991,
            4,
            22,
            15,
            0,
            0,
            0,
            ZoneId.of("Pacific/Auckland"),
        )
        val outputDocument = writingConverter.convert(inputZdt)
        assertThat<Any>(outputDocument!!["zone"]).isEqualTo("Pacific/Auckland")
        Assertions.assertThat(outputDocument.getDate("dateTime"))
            .isEqualTo(
                Date.from(
                    inputZdt.withZoneSameInstant(
                        ZoneId.of("UTC"),
                    )
                        .toInstant(),
                ),
            )
    }

    @Test
    @Throws(Exception::class)
    fun givenDocumentWithCorrectFields_whenReadingDocument_thenZonedDateTimeCorrectlyCreated() {
        val baseZdt = ZonedDateTime.of(
            1991,
            4,
            22,
            23,
            0,
            0,
            0,
            ZoneId.of("America/Argentina/Cordoba"),
        )
        val inputDocumentMap: MutableMap<String, Any> = HashMap()
        inputDocumentMap["zone"] = "America/Argentina/Cordoba"
        inputDocumentMap["dateTime"] = Date.from(
            baseZdt.withZoneSameInstant(ZoneId.of("UTC"))
                .toInstant(),
        )
        val inputDocument = Document(inputDocumentMap)
        val outputTdz = readerConverter.convert(inputDocument)
        Assertions.assertThat(outputTdz)
            .isEqualTo(baseZdt)
        Assertions.assertThat(outputTdz.toString())
            .isEqualTo("1991-04-22T23:00-04:00[America/Argentina/Cordoba]")
    }
}
